src/components/utils/src/timer.cc

```diff
@@ -44,175 +44,176 @@ CREATE_LOGGERPTR_GLOBAL(logger_, "Utils")
-  DCHECK_OR_RETURN_VOID(signal_value.sival_ptr)
-
-  timer::Timer* timer = static_cast<timer::Timer*>(signal_value.sival_ptr);
-  timer->OnTimeout();
-}
-}  // namespace timer
-
-namespace {
-const int kErrorCode = -1;
-
-itimerspec MillisecondsToItimerspec(const timer::Milliseconds miliseconds) {
-  struct itimerspec result;
-
-  result.it_value.tv_sec = miliseconds / DateTime::MILLISECONDS_IN_SECOND;
-  result.it_value.tv_nsec = (miliseconds % DateTime::MILLISECONDS_IN_SECOND) *
-                            DateTime::NANOSECONDS_IN_MILLISECOND;
-  result.it_interval.tv_sec = 0;
-  result.it_interval.tv_nsec = 0;
-
-  return result;
-}
-
-timer_t StartPosixTimer(timer::Timer& trackable,
-                        const timer::Milliseconds timeout) {
-  LOG4CXX_AUTO_TRACE(logger_);
-  timer_t internal_timer = NULL;
-
-  sigevent signal_event;
-  signal_event.sigev_notify = SIGEV_THREAD;
-  signal_event.sigev_notify_attributes = NULL;
-  signal_event.sigev_value.sival_ptr = static_cast<void*>(&trackable);
-  signal_event.sigev_notify_function = timer::HandlePosixTimer;
-
-  if (timer_create(CLOCK_REALTIME, &signal_event, &internal_timer) ==
-      kErrorCode) {
-    int error_code = errno;
-    LOG4CXX_FATAL(logger_,
-                  "Can`t create posix_timer. Error("
-                      << error_code << "): " << strerror(error_code));
-    return NULL;
-  }
-  const itimerspec itimer = MillisecondsToItimerspec(timeout);
-
-  if (timer_settime(internal_timer, 0, &itimer, NULL) == kErrorCode) {
-    int error_code = errno;
-    UNUSED(error_code);
-    LOG4CXX_FATAL(logger_,
-                  "Can`t set timeout to posix_timer. Error("
-                      << error_code << "): " << strerror(error_code));
-    return NULL;
-  }
-  return internal_timer;
-}
-
-bool StopPosixTimer(timer_t timer) {
-  LOG4CXX_AUTO_TRACE(logger_);
-  const int resultCode = timer_delete(timer);
-  if (kErrorCode == resultCode) {
-    int error_code = errno;
-    LOG4CXX_ERROR(logger_,
-                  "Can`t delete posix_timer. Error("
-                      << error_code << "): " << strerror(error_code));
-    return false;
-  }
-  return true;
-}
-}  // namespace
-
-timer::Timer::Timer(const std::string& name, const TimerTask* task_for_tracking)
+timer::Timer::Timer(const std::string& name, TimerTask* task_for_tracking)
     : name_(name)
     , task_(task_for_tracking)
-    , repeatable_(false)
-    , timeout_ms_(0u)
-    , is_running_(false)
-    , timer_(NULL) {
+    , delegate_(this)
+    , thread_(threads::CreateThread(name_.c_str(), &delegate_)) {
   LOG4CXX_AUTO_TRACE(logger_);
   DCHECK(!name_.empty());
-  DCHECK(task_);
+  DCHECK(task_.get());
+  DCHECK(thread_);
 }
 
 timer::Timer::~Timer() {
   LOG4CXX_AUTO_TRACE(logger_);
+  sync_primitives::AutoLock auto_lock(lock_);
   LOG4CXX_DEBUG(logger_, "Timer is to be destroyed " << name_);
-  Stop();
-  sync_primitives::AutoLock auto_lock(task_lock_);
-  DCHECK(task_);
-  delete task_;
+
+  DCHECK(thread_);
```

Comment:
@pestOO, I think, if we lost pointer that we creating anyway in constructor it's reason to crash with cry "all is lost!". IMHO.